hudi-test-suite/src/main/java/org/apache/hudi/testsuite/generator/DeltaGenerator.java

```diff
@@ -108,6 +114,17 @@ public DeltaGenerator(DeltaConfig deltaOutputConfig, JavaSparkContext jsc, Spark
     return inputBatch;
   }
 
+  public JavaRDD<GenericRecord> generateUpsertsWithDistributedSource(Config operation) {
```

Comment:
Not sure I follow this method properly : 
1) We shouldn't use any Test* names in the source code. 
2) How do we plan to support more properties in the future, do we need to make code changes every time ?
3) What is the need for rocksdb use here, again "Test" in the name ?
4) fetchNext is always fetching 1000000 records, why is that ?

Reply:
> Not sure I follow this method properly :
> 
> 1. We shouldn't use any Test* names in the source code.
> 2. How do we plan to support more properties in the future, do we need to make code changes every time ?
> 3. What is the need for rocksdb use here, again "Test" in the name ?
> 4. fetchNext is always fetching 1000000 records, why is that ?
> 5. The name of the method says "upsert" but I see only inserts getting generated..

Hi @n3nash When implementing this function, I also have some questions (e.g..yours 1,2,3). Actually, this is @vinothchandar 's suggestion.

I know `DistributedTestDataSource` has some limitations. IMO, we can refactor it to make it more scalability and to be a general data generator. WDYT?

About, question 4:

> 4. fetchNext is always fetching 1000000 records, why is that ?

It is because of this statement:

>     InputBatch<JavaRDD<GenericRecord>> batch = distributedTestDataSource.fetchNext(Option.empty(), 10000000);

Will try to fix it.

About, question 5:

> The name of the method says "upsert" but I see only inserts getting generated..

In `AbstractBaseTestSource#fetchNextBatch`, it connects `insertStream` and `updateStream`. IMO, it can provide upsert function. I will try to figure it out.

Reply:
@yanghua Yes, we should refactor those parts. 

For (5), what I mean is that when we perform distributedTestDataSource.fetchNext(Option.empty(), 10000000) does it return a bunch of updates + inserts (or just inserts) ?

Reply:
I have debugged the call chain of the relevant methods. 

The key chain lists below:

```
DistributedTestDataSource#fetchNext
  DistributedTestDataSource#fetchNewData
    DistributedTestDataSource#fetchNextBatch
```

In `DistributedTestDataSource#fetchNextBatch`, it will calculate the number of insert and update records. Core logic:

```
int numExistingKeys = dataGenerator.getNumExistingKeys();

int numUpdates = Math.min(numExistingKeys, sourceLimit / 2);
int numInserts = sourceLimit - numUpdates;
```

The `sourceLimit` variable is specified by the outside (here is 10000000). However, about `numExistingKeys` variable, it is always `0`. It can only be changed after calling some methods in `HoodieTestDataGenerator` to generate insert records. In our scene, these methods have never been invoked. So here:

```
numUpdates = 0;
numInserts = sourceLimit;
```

Reply:
okay, so I'm still unclear, can we pass the exact number of inserts/upserts to create using the above logic ? If not, this might not be that useful.